### PR TITLE
Set always ifindex for enlinkd links

### PR DESCRIPTION
**Pull request by rssntn67**
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13943

### CODE REVIEW

**Comment by rssntn67** [comment]
I have reviews the code to certify ifIndex is always set.
The only problem was with LLDP and I provided a method to get the ifindex so that is persisted in lldplink.
Once persisted it is automatically shown in topology.

File: features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java [hunk 1]
```diff
@@ -75,7 +77,7 @@ public List<LldpLink> findByNodeId(Integer nodeId) {
     @Override
     public void deleteByNodeIdOlderThen(Integer nodeId, Date now) {
         getHibernateTemplate().bulkUpdate("delete from LldpLink lldpLink where lldpLink.node.id = ? and lldpLink.lldpLinkLastPollTime < ?",
-                                          new Object[] {nodeId,now});
+                nodeId,now);
```
Comment (christianpape) [inline features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java:80]:
```suggestion
                nodeId, now);
```

File: features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java [hunk 2]
```diff
@@ -105,6 +107,34 @@ public List<LldpLink> findLinksForIds(List<Integer> linkIds) {
...
+                getHibernateTemplate().find("SELECT snmpIf FROM OnmsSnmpInterface AS snmpIf WHERE snmpIf.node.id = ? AND (LOWER(snmpIf.ifDescr) = LOWER(?) OR LOWER(snmpIf.ifName) = LOWER(?) OR snmpIf.physAddr = ?)",
+                       nodeid,
+                        portId,
+                       portId,
+                        portId
+                );
+        if (ifaces.size() == 1) {
+            return ((OnmsSnmpInterface) ifaces.iterator().next()).getIfIndex();
+        }
+        ifaces = getHibernateTemplate().find("SELECT ipIf From OnmsIpInterface AS ipIf where ipIf.node.id = ? and ipIf.ipAddress = ?", nodeid,portId);
```
Comment (christianpape) [inline features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java:126]:
```suggestion
        ifaces = getHibernateTemplate().find("SELECT ipIf FROM OnmsIpInterface AS ipIf WHERE ipIf.node.id = ? AND ipIf.ipAddress = ?", nodeid, portId);
```

File: features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java [hunk 3]
```diff
@@ -89,7 +91,7 @@ public List<LldpLink> findLinksForIds(List<Integer> linkIds) {
         final StringBuilder sql = new StringBuilder();
         sql.append("FROM LldpLink lldplink ");
         if(linkIds.size() == 1){
-            sql.append("where lldplink.id = " + linkIds.get(0) + " ");
+            sql.append("where lldplink.id = ").append(linkIds.get(0)).append(" ");
```
Comment (christianpape) [inline features/enlinkd/persistence/impl/src/main/java/org/opennms/netmgt/enlinkd/persistence/impl/LldpLinkDaoHibernate.java:94]:
```suggestion
            sql.append("WHERE lldplink.id = ").append(linkIds.get(0)).append(" ");
```

File: features/enlinkd/service/impl/src/main/java/org/opennms/netmgt/enlinkd/service/impl/LldpTopologyServiceImpl.java [hunk 4]
```diff
@@ -91,8 +91,12 @@ public void reconcile(int nodeId, Date now) {
     public void store(int nodeId, LldpLink link) {
         if (link == null)
             return;
+        if (link.getLldpPortIfindex() == null ) {
+            LOG.debug("store: ifindex is null, {}", link);
+            link.setLldpPortIfindex(m_lldpLinkDao.getIfIndex(nodeId,link.getLldpPortId()));
```
Comment (christianpape) [inline features/enlinkd/service/impl/src/main/java/org/opennms/netmgt/enlinkd/service/impl/LldpTopologyServiceImpl.java:96]:
```suggestion
            link.setLldpPortIfindex(m_lldpLinkDao.getIfIndex(nodeId, link.getLldpPortId()));
```

File: features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java [hunk 5]
```diff
@@ -64,13 +64,61 @@
...
+
+    }
+
+    protected static void printOnmsTopologyVertex(OnmsTopologyVertex vertex) {
+        System.err.println("OnmsTopologyVertex: id:"+vertex.getId());
+        System.err.println("OnmsTopologyVertex: nodeid:"+vertex.getNodeid());
+        System.err.println("OnmsTopologyVertex: label:"+vertex.getLabel());
+        System.err.println("OnmsTopologyVertex: address:"+vertex.getAddress());
+        System.err.println("OnmsTopologyVertex: iconKey:"+vertex.getIconKey());
+        System.err.println("OnmsTopologyVertex: toolTipText:"+vertex.getToolTipText());
```
Comment (christianpape) [inline features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java:118]:
```suggestion
        System.err.println("OnmsTopologyVertex: id:" + vertex.getId());
        System.err.println("OnmsTopologyVertex: nodeid:" + vertex.getNodeid());
        System.err.println("OnmsTopologyVertex: label:" + vertex.getLabel());
        System.err.println("OnmsTopologyVertex: address:" + vertex.getAddress());
        System.err.println("OnmsTopologyVertex: iconKey:" + vertex.getIconKey());
        System.err.println("OnmsTopologyVertex: toolTipText:" + vertex.getToolTipText());
```

File: features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java [hunk 6]
```diff
@@ -64,13 +64,61 @@
...
+
+        System.err.println("OnmsTopologyEdge: target/label:" + edge.getTarget().getVertex().getLabel());
+        System.err.println("OnmsTopologyEdge: target/index:" + edge.getTarget().getIndex());
+        System.err.println("OnmsTopologyEdge: target/ifindex:" + edge.getTarget().getIfindex());
+        System.err.println("OnmsTopologyEdge: target/ifname:" + edge.getTarget().getIfname());
+        System.err.println("OnmsTopologyEdge: target/addr:" + edge.getTarget().getAddr());
+        System.err.println("OnmsTopologyEdge: target/speed:" + edge.getTarget().getSpeed());
+        System.err.println("OnmsTopologyEdge: target/toolTipText:" + edge.getTarget().getToolTipText());
+
+
```
Comment (christianpape) [inline features/enlinkd/tests/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdTestHelper.java:109]: remove empty lines?

**Review by christianpape — CHANGES_REQUESTED** [review]
Only minor changes needed

**Review by christianpape — APPROVED** [review]
LGTM.

**Pull request merged**